be/src/vec/exec/format/parquet/parquet_column_convert.h

```diff
@@ -0,0 +1,623 @@
+            if (!timestamp_info.isAdjustedToUTC) {
+                // should set timezone to utc+0
+                ctz = const_cast<cctz::time_zone*>(&utc0);
+            }
+            const auto& time_unit = timestamp_info.unit;
+            if (time_unit.__isset.MILLIS) {
+                second_mask = 1000;
+                scale_to_nano_factor = 1000000;
+            } else if (time_unit.__isset.MICROS) {
+                second_mask = 1000000;
+                scale_to_nano_factor = 1000;
+            } else if (time_unit.__isset.NANOS) {
+                second_mask = 1000000000;
+                scale_to_nano_factor = 1;
+            }
+        } else if (schema.__isset.converted_type) {
+            const auto& converted_type = schema.converted_type;
+            if (converted_type == tparquet::ConvertedType::TIMESTAMP_MILLIS) {
+                second_mask = 1000;
+                scale_to_nano_factor = 1000000;
+            } else if (converted_type == tparquet::ConvertedType::TIMESTAMP_MICROS) {
+                second_mask = 1000000;
+                scale_to_nano_factor = 1000;
+            }
+        }
+
+        if (ctz) {
+            VecDateTimeValue t;
+            t.from_unixtime(0, *ctz);
+            offset_days = t.day() == 31 ? 0 : 1;
+        }
+    }
+
+    template <typename DecimalPrimitiveType>
+    void init_decimal_converter(DataTypePtr& data_type) {
+        if (field_schema == nullptr || decimal_scale.scale_type != DecimalScaleParams::NOT_INIT) {
+            return;
+        }
+        auto scale = field_schema->parquet_schema.scale;
+        auto* decimal_type = static_cast<DataTypeDecimal<Decimal<DecimalPrimitiveType>>*>(
+                const_cast<IDataType*>(remove_nullable(data_type).get()));
+        auto dest_scale = decimal_type->get_scale();
+        if (dest_scale > scale) {
+            decimal_scale.scale_type = DecimalScaleParams::SCALE_UP;
+            decimal_scale.scale_factor =
+                    DecimalScaleParams::get_scale_factor<DecimalPrimitiveType>(dest_scale - scale);
+        } else if (dest_scale < scale) {
+            decimal_scale.scale_type = DecimalScaleParams::SCALE_DOWN;
+            decimal_scale.scale_factor =
+                    DecimalScaleParams::get_scale_factor<DecimalPrimitiveType>(scale - dest_scale);
+        } else {
+            decimal_scale.scale_type = DecimalScaleParams::NO_SCALE;
+            decimal_scale.scale_factor = 1;
+        }
+    }
+};
+
+Status convert_data_type_from_parquet(tparquet::Type::type parquet_type, PrimitiveType,
+                                      vectorized::DataTypePtr& ans_data_type, DataTypePtr& src_type,
+                                      bool* need_convert);
+
+struct ColumnConvert {
+    virtual Status convert(const IColumn* src_col, IColumn* dst_col) { return Status::OK(); }
+
+    virtual ~ColumnConvert() = default;
+
+    void convert_null(const IColumn** src_col, IColumn** dst_col) const {
```

Comment:
Why use double `**`(pointer of pointer)? We'd better use `IColumn& col` as the parameter, or `MutableColumnPtr& col` to stand the column to be altered.